docs/proposals/107-spark-client/README.md

```diff
@@ -0,0 +1,597 @@
+```
+
+**Backend**: KubernetesBackend using Spark Operator CRDs
+
+- Extensible for future backends (Gateway/Livy)
+- Batch job support can be added later
+
+---
+
+## Backend Architecture
+
+The SparkClient uses a pluggable backend architecture that supports both direct Kubernetes access and REST API-based services:
+
+```
+                        SparkBackend (ABC)
+                              |
+              +---------------+---------------+
+              |                               |
+      KubernetesBackend              RESTSparkBackend (ABC)
+      - SparkConnect CRD                      |
+      - SparkApplication CRD        +---------+---------+
+                                    |                   |
+                            GatewayBackend        LivyBackend
+                            - BPG REST API        - Livy REST API
+                            - Queue routing       - Session mgmt
+                            - Multi-cluster       - Batch/Interactive
+```
+
+### Backend Implementations
+
+| Backend | Description | Use Case |
+|---------|-------------|----------|
+| **KubernetesBackend** | Direct K8s API with Spark Operator CRDs | Default, single cluster |
+| **GatewayBackend** | Batch Processing Gateway REST API | Multi-cluster, queue routing |
+| **LivyBackend** | Apache Livy REST API | Legacy systems, YARN integration |
+
+### Selecting a Backend
+
+```python
+from kubeflow.spark import SparkClient
+from kubeflow.spark.backends import KubernetesBackendConfig, GatewayBackendConfig
+
+# Default: Kubernetes backend
+client = SparkClient.builder().build()
+
+# Gateway backend for multi-cluster
+client = SparkClient.builder().backend(
+    GatewayBackendConfig(
+        base_url="https://gateway.example.com",
+        queue="production",
+    )
+).build()
+```
+
+---
+
+## Kubeflow Ecosystem Integration
+
+SparkClient integrates with the broader Kubeflow ecosystem:
+
+```
+SparkClient
+    |
+    +---> Kubeflow Pipelines (SparkJobOp)
+    |         - Pipeline step for Spark ETL
+    |         - DAG orchestration with Spark jobs
+    |
+    +---> Kubeflow Trainer
+    |         - Feature preparation -> Training workflow
+    |         - S3DatasetInitializer with Spark output
+    |
+    +---> Spark History MCP Server
+              - AI-powered job analysis
+              - Performance bottleneck detection
+              - Query job metrics via natural language
+```
+
+### Integration with Kubeflow Pipelines
+
+```python
+from kfp import dsl
+from kubeflow.spark.pipelines import SparkJobOp
+
+@dsl.pipeline(name="ml-pipeline")
+def ml_pipeline():
+    # Spark ETL step
+    etl = SparkJobOp(
+        name="feature-etl",
+        main_file="s3a://ml/etl.py",
+        executor_instances=20,
+        executor_memory="8g",
+    )
+    
+    # Training step depends on ETL completion
+    train = TrainOp(
+        dataset_path=etl.outputs["output_path"],
+    )
+    train.after(etl)
```

Comment:
We should whether we need to move it under: `kubeflow.pipelines.components`.
We can also leverage KFP lightweight Python tasks to submit such pipelines.

cc @franciscojavierarceo @kubeflow/wg-pipeline-leads